### DISCUSSION TITLE
How to generalize the GUI code

### DISCUSSION THREAD

**Discussion by almarklein** [pygfx/pygfx#141] [Ideas]
Let's try a GH discussion!

I think @Korijn's #140 tries to do three things:
* Make it possible to write an example that is GUI agnostic (and with less GUI boilerplate).
* Reduce some boilerplate by automatically creating a camera and controls.
* For later, introduce features like getting info as you pick objects in the scene.

I'd like to focus on that first point here. This is going somewhat in the direction of how Vispy backends work. I do agree that it's a good idea to abstract away the GUI details; it's not great that all our examples use (a specific version of) Qt. So here are some thoughts.

### Events

In Vispy, there is a custom event system, and the GUI backends convert the native events to this system. I *don't* want to do that. Instead, I'd propose that in `wgpu-py`'s GUI backends, we make events go through one method, which can then be subclassed (in pygfx) to handle events in a generic manner. I'm basically proposing [how jupyter_rfb does it](https://jupyter-rfb.readthedocs.io/en/latest/reference.html#jupyter_rfb.RemoteFrameBuffer.handle_event).

In other words, if wgpu-py provides a generic way to handle events, then in Pygfx we can very easily put the control on top.

### Event loop

Similarly, it looks like we need a generalization for starting the event loop for a GUI backend. Maybe this should also be provided by wgpu-py.

### Timers

Since the event loop differs, async timers (ala "call this function 1 second from now") are also GUI specific. If only all event loops were based on asyncio ...

### wgpu-py?

If we go down this route, the backends in wgpu-py do get bigger. I wonder if that justifies putting these in a separate package.

**Comment by Korijn** [comment]
I like what you are proposing, it makes sense to me. Abstracting away all those parts would indeed allow the viewer to be "agnostic" of the gui layer that is being used.

I do want to point out that the main intention of pygfx is to provide a library, and not to provide a viewer. The viewer utility I put together is kind of an "extra", even though I expect it will be seen as the "main entrypoint" by a large portion of potential users.

One objection I have to putting all this stuff in wgpu-py, is that we are fulfilling needs of a utility in pygfx, not wgpu-py. In other words, I am not sure you can say these abstractions would serve a purpose as part of wgpu-py if you squint and pretend pygfx doesn't exist for a moment. So it feels kind of weird to put it in wgpu-py, from that perspective. Another way to put it; we're only going to get more reasons from pygfx to change the gui abstraction, not from wgpu-py. So I feel like it only makes the development flow more cumbersome, really.

(Just thinking out loud here, not offering any solutions yet...)

**Reply by almarklein** [reply]
> The viewer utility I put together is kind of an "extra", even though I expect it will be seen as the "main entrypoint"

I agree :)

> One objection I have to putting all this stuff in wgpu-py, is that we are fulfilling needs of a utility in pygfx, not wgpu-py

Very true, but I think that other apps/libs using wgpu-py are likely to benefit from such abstractions. This means it would broaden the scope of what wgpu-py offers. This is what I meant with my last comment in the first post - perhaps this should be a completely different library.

**Reply by Korijn** [reply]
I can see what you mean now. Basically it's a GUI abstraction layer that tries to keep it simple and provide generic window creation and input handling functionality. Kind of similar to what wgpu does for graphics APIs!

I think it's a cool idea but definitely much more work than I wanted to take on right now ☺️

I also have trouble imagining how jupyter would fit into this picture. Maybe it just doesn't.

**Reply by almarklein** [reply]
> I also have trouble imagining how jupyter would fit into this picture. Maybe it just doesn't.

The Jupyter_rfb backend would follow the same approach. Except that I'd propose to let all backends process events the way jupyter_rfb does.

**Comment by almarklein** [comment]
Another solution that @berendkleinhaneveld and I briefly discussed last week is to only implement that generic event abstraction for glfw. Then glfw and jupyter_rfb have the same API for events, and both work with asyncio (glfw has no event loop, so we can just use asyncio). That way, our examples work both "on desktop" and in Jupyter (without having to change the code).

Basically, the glfw and jupyter_rfb backends are for simpler applications and examples, and for these hooking up controllers is easier. The other backends are for when you want to include pygfx in an application, for these you need a bit of extra code to hook up the controllers to the GUI's event system.

**Reply by almarklein** [reply]
One BUT. It looks like glfw does not have macos_arm64 wheels yet.

**Reply by almarklein** [reply]
Although glfw should still work if you bring your own libglfw: https://github.com/FlorianRhiem/pyGLFW/pull/58

**Reply by berendkleinhaneveld** [reply]
That's just a matter of time, right? :)

**Reply by almarklein** [reply]
Yes. With wgpu-native we're lucky because it's Rust, and Rust is good at crosscompiling: the macos_arm64 binaries are build on an intel mac VM. Doing that for C libraries is much harder (from what I understand).

**Reply by almarklein** [reply]
Not sure what the status is of "M1 runners" on CI. E.g. how does numpy builds these bins?

**Reply by Korijn** [reply]
All the cool kids are migrating to cibuildwheel which can also do cross compilation for mixed python/c projects.

We use it too now for cython, c++ and numba projects. It's really nicely designed.

**Reply by almarklein** [reply]
Oh wow. I think you've mentioned that before. That sounds great. Maybe we can help pyglfw move to that.

**Reply by almarklein** [reply]
Euhm, can cffi also not use that? :D

**Reply by Korijn** [reply]
They probably should!
